File: day04/src/main/kotlin/Rules.kt
fun String.inRange(min:Int, max:Int):Boolean = toIntOrNull()?.let{ it in min..max } ?: false


val byrRule = { credential:String -> credential.inRange(1920,2002)    }
val iyrRule = { credential:String -> credential.inRange(2010, 2020) }
val eyrRule = { credential:String -> credential.inRange(2020, 2030)  }
val hgtRule = { credential:String ->
           (credential.endsWith("cm") && credential.removeSuffix("cm").inRange(150, 193))
        || (credential.endsWith("in") && credential.removeSuffix("in").inRange(59, 76))
}
val hclRule = { credential:String ->
               credential.length == 7
            && credential.startsWith("#")
            && credential.drop(1).all{"0123456789abcdef".contains(it)}
}
val eclRule = { credential:String -> listOf("amb", "blu", "brn", "gry", "grn", "hzl", "oth").contains(credential)}
val pidRule = { credential:String ->  (credential.length == 9)  && (credential.all{"0123456789".contains(it)})}
val cidRule = { credential:String -> true}

File: day04Swift/Sources/day04Swift/rules.swift
//
//  File.swift
//  
//
//  Created by Michael Neilens on 04/12/2020.
//
import Foundation

extension String {
    func inRange(_ range:ClosedRange<Int>) -> Bool {
        guard let value =  Int(self) else { return false}
        return range.contains(value)
    }
}

func byrRule(_ credential:String) -> Bool  { credential.inRange(1920...2002) }
func iyrRule(_ credential:String) -> Bool  { credential.inRange(2010...2020) }
func eyrRule(_ credential:String) -> Bool  { credential.inRange(2020...2030) }

func hgtRule(_ credential:String) -> Bool  {
    credential.hasSuffix("cm") && String(credential.dropLast(2)).inRange(150...193)
    || credential.hasSuffix("in") && String(credential.dropLast(2)).inRange(59...76)
}

func hclRule(_ credential:String) -> Bool  {
    return credential.count == 7 && credential.hasPrefix("#") && !credential.map{ "0123456789abcdef".contains($0)}.dropFirst().contains(false)
}

func eclRule(_ credential:String) -> Bool  {["amb", "blu", "brn", "gry", "grn", "hzl", "oth"].contains(credential)}
func pidRule(_ credential:String) -> Bool  {credential.count == 9 && !credential.map{ "0123456789".contains($0)}.contains(false) }
func cidRule(_ credential:String) -> Bool  {true}
